### DISCUSSION TITLE
Any example on how to render custom webgl scene into a window

### DISCUSSION THREAD

**Discussion by shi-yan** [emilk/egui#1172] [Q&A]
I want to render a 3D scene into a window. My main target is the web.

With imgui, we usually need to get a window's content rectangle, and we render the scene into a texture and apply to the window's content area.

Any example how to achieve this with egui?

**Comment by emilk** [comment]
There are examples for how to do this for `egui-miniquad`: here: https://github.com/not-fl3/egui-miniquad/blob/master/examples/render_to_egui_image.rs (and it can be compiled for web).

There are currently no example for how to do this using `egui_web`.

If you are rendering something complex in 3D you may perhaps want to use a more high-level 3D library, e.g. using `bevy_egui` or `egui-macroquad` (both which can be compiled to web).

**Reply by shi-yan** [reply]
Thank you for the links, I still have some questions.

Is it possible to do this using egui_web? Or I have toswitch to bevy_egui or egui-macroquad.

if egui_web is capable to do this, how can I obtain the webgl context through egui_web::ctx_Ref or the egui_web::frame struct?

or how can I get the webgl context somewhere else? Do I need to obtain the context myself using the canvas' id?

**Reply by emilk** [reply]
`egui_web` has so far been designed to be used as an `eframe` backend, so it will likely need a bit of a refactor to be useable stand-alone.

Since we're moving towards having `glow` as the default painter for `eframe` (on both web and native), we could consider adding an (optional) glow context reference to `epi::Frame` for this sort of work. It would be very nice if one could just do some OpenGL painting in ones own `App::update` function. I created an issue for this here: https://github.com/emilk/egui/issues/1198

**Reply by shi-yan** [reply]
Thank you. I was able to load texture using frame::alloc_texture. yes, I also found the current API is a little bit limiting. frame::alloc_texture doesn't accept the raw texture format (WebGLTexture in webgl case). I have to provide an epi::Image first. This is fine if there is an image. But this is inconvenient for the framebuffer case.

I hope frame can 

1. provide the raw context
2. provide a function to create framebuffer
3. when a window/widget 's size has changed, auto resize the framebuffer?
